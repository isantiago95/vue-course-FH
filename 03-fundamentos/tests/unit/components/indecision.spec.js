import Indecision from '@/components/Indecision';
import { shallowMount } from '@vue/test-utils';

describe('indecision component', () => {

    let wrapper;
    let clgSpy;

    globalThis.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: "yes",
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif"
        })
    }));

    beforeEach(() => {
        wrapper = shallowMount(Indecision);

        clgSpy = jest.spyOn(console, 'log');
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('should match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('should not dispatch anything when type (console.log)', async () => {
        const getAnswerSpy = jest.spyOn(Indecision.methods, 'getAnswer');

        const input = wrapper.find('input');
        await input.setValue('Hola mundo');

        expect(clgSpy).toHaveBeenCalledTimes(1);
        expect(getAnswerSpy).not.toHaveBeenCalled();
    });

    test('should dispatch getAnswer when "?" is typed', async () => {
        const getAnswerSpy = jest.spyOn(Indecision.methods, 'getAnswer');

        const input = wrapper.find('input');
        await input.setValue('Hola mundo?');

        expect(clgSpy).toHaveBeenCalledTimes(1);
        expect(getAnswerSpy).toHaveBeenCalledTimes(1);
    });

    test('should getAnswer() work', async () => {

        await wrapper.vm.getAnswer();

        const img = wrapper.find('img');
        expect(img.exists()).toBeTruthy();
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif');
        expect(wrapper.vm.answer).toBe('Si!');
    });

    test('should getAnswer() failed', async () => {

        fetch.mockImplementationOnce(() => Promise.reject('API is down'));

        await wrapper.vm.getAnswer();

        const img = wrapper.find('img');
        expect(img.exists()).toBeFalsy();
        expect(wrapper.vm.answer).toBe('No se pudo conectar con el servidor');
    });
});