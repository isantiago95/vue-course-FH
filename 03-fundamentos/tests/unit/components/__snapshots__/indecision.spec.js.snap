// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indecision component should match with snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de interrogacion (?)</p>
  <!--v-if-->
</div>
`;
